File: apps/mobile/plugins/remote/ios/Sources/VelaRemotePlugin/ConnectionRecords.swift
import Foundation

/// Connection identity includes credentials and pairing fragments, never presentation or tunnel state.
enum ConnectionRecords {
    static let secrets = ["password", "privateKey", "passphrase", "webPassword"]
    static func identity(_ row: [String: Any]) -> [String] {
        func s(_ key: String) -> String { row[key] as? String ?? "" }
        if s("mode") == "url" { return ["url", s("url"), s("webPassword")] }
        return [s("mode"), s("host").lowercased(), String((row["port"] as? NSNumber)?.intValue ?? 0), s("username"), s("auth"),
                s("auth") == "key" ? s("privateKey") : s("password"), s("auth") == "key" ? s("passphrase") : "",
                s("service"), s("service") == "manual" ? String((row["remotePort"] as? NSNumber)?.intValue ?? 0) : "",
                s("service") == "auto" ? String(row["prepare"] as? Bool ?? false) : "", s("webPassword")]
    }
    static func prepared(_ input: [String: Any], in rows: [[String: Any]]) -> [String: Any] {
        var row = input
        var old = rows.first { ($0["id"] as? String) == (input["id"] as? String) }
        // Re-scanning a saved address does not ask the user to re-enter an existing secret.
        if old == nil, (input["id"] as? String ?? "").isEmpty, secrets.allSatisfy({ (input[$0] as? String ?? "").isEmpty }) {
            func withoutSecrets(_ value: [String: Any]) -> [String] {
                var value = value; for key in secrets { value.removeValue(forKey: key) }; return identity(value)
            }
            let matches = rows.filter { withoutSecrets($0) == withoutSecrets(input) }
            if matches.count == 1 { old = matches[0]; for key in secrets { row.removeValue(forKey: key) } }
        }
        for key in secrets where row[key] == nil { row[key] = old?[key] }
        row["id"] = old?["id"] ?? UUID().uuidString
        row["localPort"] = old?["localPort"] ?? 0
        return row
    }
    static func copied(_ input: [String: Any], from source: [String: Any]) -> [String: Any] {
        var row = input
        for key in secrets where row[key] == nil { row[key] = source[key] }
        row["id"] = UUID().uuidString
        row["localPort"] = 0
        return row
    }
    static func upsert(_ input: [String: Any], into rows: inout [[String: Any]], preserveEquivalent: Bool = false) -> [String: Any] {
        var row = input
        let same = rows.first { identity($0) == identity(row) }
        // Saving an unchanged copy must not rename or otherwise overwrite its source.
        if preserveEquivalent, let same { return same }
        let existing = rows.first { ($0["id"] as? String) == (row["id"] as? String) }
        if existing == nil, let same { row["id"] = same["id"]; row["localPort"] = same["localPort"] }
        let index = rows.firstIndex { ($0["id"] as? String) == (row["id"] as? String) } ?? rows.count
        rows.removeAll { ($0["id"] as? String) == (row["id"] as? String) || identity($0) == identity(row) }
        rows.insert(row, at: min(index, rows.count))
        return row
    }
    static func unique(_ rows: [[String: Any]]) -> [[String: Any]] {
        var result = [[String: Any]]()
        for row in rows where !result.contains(where: { identity($0) == identity(row) }) { result.append(row) }
        return result
    }
    static func summary(_ row: [String: Any]) -> [String: Any] {
        var result = row
        result["hasSecret"] = !(row["password"] as? String ?? "").isEmpty || !(row["privateKey"] as? String ?? "").isEmpty
        result["hasWebPassword"] = !(row["webPassword"] as? String ?? "").isEmpty
        for key in secrets { result.removeValue(forKey: key) }
        return result
    }
}

File: apps/mobile/plugins/remote/android/src/main/java/com/velaterm/remote/ConnectionRecords.kt
package com.velaterm.remote

import org.json.JSONArray
import org.json.JSONObject
import java.util.Locale
import java.util.UUID

/** Identity includes credentials and pairing fragments, never presentation or tunnel state. */
internal object ConnectionRecords {
    val secrets = listOf("password", "privateKey", "passphrase", "webPassword")
    fun identity(row: JSONObject): List<String> {
        fun s(key: String) = row.optString(key, "")
        if (s("mode") == "url") return listOf("url", s("url"), s("webPassword"))
        return listOf(s("mode"), s("host").lowercase(Locale.ROOT), row.optInt("port", 0).toString(), s("username"), s("auth"),
            if (s("auth") == "key") s("privateKey") else s("password"), if (s("auth") == "key") s("passphrase") else "",
            s("service"), if (s("service") == "manual") row.optInt("remotePort", 0).toString() else "",
            if (s("service") == "auto") row.optBoolean("prepare", false).toString() else "", s("webPassword"))
    }
    fun prepared(input: JSONObject, rows: JSONArray): JSONObject {
        val row = JSONObject(input.toString())
        val items = (0 until rows.length()).map { rows.getJSONObject(it) }
        var old = items.firstOrNull { it.optString("id") == input.optString("id") }
        // Re-scanning a saved address preserves its secret when there is one unambiguous match.
        if (old == null && input.optString("id").isEmpty() && secrets.all { input.optString(it, "").isEmpty() }) {
            fun withoutSecrets(value: JSONObject) = identity(JSONObject(value.toString()).apply { secrets.forEach { remove(it) } })
            val matches = items.filter { withoutSecrets(it) == withoutSecrets(input) }
            if (matches.size == 1) { old = matches[0]; secrets.forEach { row.remove(it) } }
        }
        for (key in secrets) if (!row.has(key) && old != null) row.put(key, old.optString(key, ""))
        row.put("id", old?.getString("id") ?: UUID.randomUUID().toString())
        row.put("localPort", old?.optInt("localPort", 0) ?: 0)
        return row
    }
    fun copied(input: JSONObject, source: JSONObject): JSONObject = JSONObject(input.toString()).apply {
        for (key in secrets) if (!has(key)) put(key, source.optString(key, ""))
        put("id", UUID.randomUUID().toString())
        put("localPort", 0)
    }
    fun upsert(input: JSONObject, rows: JSONArray, preserveEquivalent: Boolean = false): Pair<JSONObject, JSONArray> {
        val row = JSONObject(input.toString())
        val items = (0 until rows.length()).map { rows.getJSONObject(it) }
        val same = items.firstOrNull { identity(it) == identity(row) }
        // Saving an unchanged copy must not rename or otherwise overwrite its source.
        if (preserveEquivalent && same != null) return same to rows
        if (items.none { it.optString("id") == row.optString("id") } && same != null) {
            row.put("id", same.getString("id")); row.put("localPort", same.optInt("localPort", 0))
        }
        val index = items.indexOfFirst { it.optString("id") == row.optString("id") }.let { if (it < 0) items.size else it }
        val result = items.filterNot { it.optString("id") == row.optString("id") || identity(it) == identity(row) }.toMutableList()
        result.add(minOf(index, result.size), row)
        return row to JSONArray(result)
    }
    fun unique(rows: JSONArray) = JSONArray((0 until rows.length()).map { rows.getJSONObject(it) }.distinctBy { identity(it) })
    fun summary(row: JSONObject): JSONObject = JSONObject(row.toString()).apply {
        put("hasSecret", optString("password").isNotEmpty() || optString("privateKey").isNotEmpty())
        put("hasWebPassword", optString("webPassword").isNotEmpty())
        secrets.forEach { remove(it) }
    }
}
